RFS/src/main/java/com/rfs/common/LuceneDocumentsReader.java

```diff
@@ -11,40 +10,65 @@
 import org.apache.lucene.index.IndexReader;
 import org.apache.lucene.store.FSDirectory;
 import org.apache.lucene.util.BytesRef;
+import reactor.core.publisher.Flux;
 
 public class LuceneDocumentsReader {
     private static final Logger logger = LogManager.getLogger(LuceneDocumentsReader.class);
 
-    public static List<Document> readDocuments(Path luceneFilesBasePath, String indexName, int shardId) throws Exception {
+    public Flux<Document> readDocuments(Path luceneFilesBasePath, String indexName, int shardId) {
         Path indexDirectoryPath = luceneFilesBasePath.resolve(indexName).resolve(String.valueOf(shardId));
 
-        List<Document> documents = new ArrayList<>();
+        return Flux.using(
+            () -> openIndexReader(indexDirectoryPath),
+            reader -> {
+                logger.info(reader.maxDoc() + " documents found in the current Lucene index");
 
-        try (FSDirectory directory = FSDirectory.open(indexDirectoryPath);
-             IndexReader reader = DirectoryReader.open(directory)) {
-
-            // Add all documents to our output that have the _source field set and filled in
-            for (int i = 0; i < reader.maxDoc(); i++) {
-                Document document = reader.document(i);
-                BytesRef source_bytes = document.getBinaryValue("_source");
-                String id;
-                // TODO Improve handling of missing document id (https://opensearch.atlassian.net/browse/MIGRATIONS-1649)
+                return Flux.range(0, reader.maxDoc()) // Extract all the Documents in the IndexReader
+                .handle((i, sink) -> {
+                    Document doc = getDocument(reader, i);
+                    if (doc != null) { // Skip malformed docs
```

Comment:
You should at least log when doc == null (or whatever malformed documents that you might be skipping).

Reply:
We log that in `getDocument()`